From a510cba94a51b79230b2731cda38100108b6f825 Mon Sep 17 00:00:00 2001
From: Jinjiang Tu <tujinjiang@huawei.com>
Date: Mon, 10 Mar 2025 10:45:08 +0800
Subject: [PATCH 889/960] downstream: mm/pfn_range_alloc: add new allocator for
 memory lending

Introduce pfn_range_{alloc,free} and hugetlb_pool_{alloc,free} for
memory lending.
It handles address mapping, cache attributes and related operations.

Signed-off-by: Jinjiang Tu <tujinjiang@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 .../admin-guide/kernel-parameters.txt         |   8 +
 arch/arm64/Kconfig                            |   1 +
 arch/arm64/mm/Kconfig                         |  15 +
 arch/arm64/mm/Makefile                        |   1 +
 arch/arm64/mm/init.c                          |   6 +
 arch/arm64/mm/internal.h                      |  29 +
 arch/arm64/mm/mmu.c                           |  26 +-
 arch/arm64/mm/pfn_range_alloc.c               | 624 ++++++++++++++++++
 include/linux/hugetlb.h                       |  20 +
 include/linux/mm.h                            |  32 +
 include/trace/events/kmem.h                   | 118 ++++
 mm/hugetlb.c                                  |  81 +++
 mm/page_alloc.c                               |  63 +-
 13 files changed, 1015 insertions(+), 9 deletions(-)
 create mode 100644 arch/arm64/mm/Kconfig
 create mode 100644 arch/arm64/mm/internal.h
 create mode 100644 arch/arm64/mm/pfn_range_alloc.c

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 1ce6dc4bcd7b..56fd5463e6c3 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -4617,6 +4617,14 @@
 	pm_debug_messages	[SUSPEND,KNL]
 			Enable suspend/resume debug messages during boot up.
 
+	pmd_mapping=	[ARM64,KNL]
+			Format: nn%
+			Allows to allocate contiguous memory from special pfn
+			range, the linear mapping granule of this range is never
+			larger than PMD. pmd_mapping specifies the percent of
+			memory of each node. pmd_mapping=100% is used for hugetlb
+			scenarios, the whole linear mapping isn't large than PMD.
+
 	pnp.debug=1	[PNP]
 			Enable PNP debug messages (depends on the
 			CONFIG_PNP_DEBUG_MESSAGES option).  Change at run-time
diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 289caf41b926..7c4172794124 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -2559,3 +2559,4 @@ source "drivers/acpi/Kconfig"
 
 source "arch/arm64/kvm/Kconfig"
 
+source "arch/arm64/mm/Kconfig"
diff --git a/arch/arm64/mm/Kconfig b/arch/arm64/mm/Kconfig
new file mode 100644
index 000000000000..e7880c691822
--- /dev/null
+++ b/arch/arm64/mm/Kconfig
@@ -0,0 +1,15 @@
+# SPDX-License-Identifier: GPL-2.0
+#
+# MEMORY configuration for arm64
+#
+
+config PFN_RANGE_ALLOC
+	bool "Enable contiguous pfn range allocator"
+	default n
+	select CONTIG_ALLOC
+	depends on MEMORY_HOTPLUG
+	help
+		It allows to allocate contiguous memory from special pfn range,
+		the linear mapping granule of this range is never larger than PMD.
+
+		If unsure, say N.
diff --git a/arch/arm64/mm/Makefile b/arch/arm64/mm/Makefile
index dbd1bc95967d..0bf43bc4e555 100644
--- a/arch/arm64/mm/Makefile
+++ b/arch/arm64/mm/Makefile
@@ -14,3 +14,4 @@ KASAN_SANITIZE_physaddr.o	+= n
 
 obj-$(CONFIG_KASAN)		+= kasan_init.o
 KASAN_SANITIZE_kasan_init.o	:= n
+obj-$(CONFIG_PFN_RANGE_ALLOC)	+= pfn_range_alloc.o
diff --git a/arch/arm64/mm/init.c b/arch/arm64/mm/init.c
index 8a0f8604348b..3094cd3d08fc 100644
--- a/arch/arm64/mm/init.c
+++ b/arch/arm64/mm/init.c
@@ -31,6 +31,7 @@
 #include <linux/hugetlb.h>
 #include <linux/acpi_iort.h>
 #include <linux/kmemleak.h>
+#include "internal.h"
 
 #include <asm/boot.h>
 #include <asm/fixmap.h>
@@ -498,6 +499,11 @@ void __init mem_init(void)
 
 	swiotlb_init(swiotlb, SWIOTLB_VERBOSE);
 
+	/* Must be placed before buddy is initialized, to avoid reserved
+	 * memory is reserved by memblock.
+	 */
+	pmd_mapping_reserve_and_remap();
+
 	/* this will put all unused low memory onto the freelists */
 	memblock_free_all();
 
diff --git a/arch/arm64/mm/internal.h b/arch/arm64/mm/internal.h
new file mode 100644
index 000000000000..6a2a4746e954
--- /dev/null
+++ b/arch/arm64/mm/internal.h
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#ifndef __ARM64_MM_INTERNAL_H
+#define __ARM64_MM_INTERNAL_H
+
+#ifdef CONFIG_PFN_RANGE_ALLOC
+#define PFN_RANGE_ALLOC_SIZE PMD_SIZE
+#define PFN_RANGE_ALLOC_ORDER PMD_ORDER
+
+static inline bool should_pmd_linear_mapping(void)
+{
+	return contig_mem_pool_percent == 100;
+}
+
+void __init pmd_mapping_reserve_and_remap(void);
+void __init pmd_mapping_reserved_remap(phys_addr_t start, phys_addr_t end);
+#else
+static inline void pmd_mapping_reserve_and_remap(void)
+{
+}
+static inline void pmd_mapping_reserved_remap(phys_addr_t start, phys_addr_t end)
+{
+}
+static inline bool should_pmd_linear_mapping(void)
+{
+	return false;
+}
+#endif
+#endif /* ifndef _ARM64_MM_INTERNAL_H */
diff --git a/arch/arm64/mm/mmu.c b/arch/arm64/mm/mmu.c
index 98601b7aeadb..47dc1603545c 100644
--- a/arch/arm64/mm/mmu.c
+++ b/arch/arm64/mm/mmu.c
@@ -25,6 +25,7 @@
 #include <linux/vmalloc.h>
 #include <linux/set_memory.h>
 #include <linux/kfence.h>
+#include "internal.h"
 
 #include <asm/barrier.h>
 #include <asm/cputype.h>
@@ -44,6 +45,11 @@
 #define NO_BLOCK_MAPPINGS	BIT(0)
 #define NO_CONT_MAPPINGS	BIT(1)
 #define NO_EXEC_MAPPINGS	BIT(2)	/* assumes FEAT_HPDS is not used */
+#ifdef CONFIG_PFN_RANGE_ALLOC
+#define NO_PUD_BLOCK_MAPPINGS BIT(3)
+#else
+#define NO_PUD_BLOCK_MAPPINGS 0
+#endif
 
 int idmap_t0sz __ro_after_init;
 
@@ -344,7 +350,7 @@ static void alloc_init_pud(pgd_t *pgdp, unsigned long addr, unsigned long end,
 		 */
 		if (pud_sect_supported() &&
 		   ((addr | next | phys) & ~PUD_MASK) == 0 &&
-		    (flags & NO_BLOCK_MAPPINGS) == 0) {
+		    (flags & (NO_BLOCK_MAPPINGS | NO_PUD_BLOCK_MAPPINGS)) == 0) {
 			pud_set_huge(pudp, phys, prot);
 
 			/*
@@ -586,6 +592,8 @@ static void __init map_mem(pgd_t *pgdp)
 
 	if (can_set_direct_map())
 		flags |= NO_BLOCK_MAPPINGS | NO_CONT_MAPPINGS;
+	else if (should_pmd_linear_mapping())
+		flags |= NO_PUD_BLOCK_MAPPINGS | NO_CONT_MAPPINGS;
 
 	/*
 	 * Take care not to create a writable alias for the
@@ -1133,6 +1141,20 @@ static void free_empty_tables(unsigned long addr, unsigned long end,
 }
 #endif
 
+#ifdef CONFIG_PFN_RANGE_ALLOC
+void __init pmd_mapping_reserved_remap(phys_addr_t start, phys_addr_t end)
+{
+	unsigned long vstart, vend;
+
+	vstart = __phys_to_virt(start);
+	vend = __phys_to_virt(end);
+	unmap_hotplug_range(vstart, vend, false, NULL);
+	__create_pgd_mapping(swapper_pg_dir, start, vstart, end - start,
+			pgprot_tagged(PAGE_KERNEL), early_pgtable_alloc,
+			NO_PUD_BLOCK_MAPPINGS | NO_CONT_MAPPINGS);
+}
+#endif
+
 void __meminit vmemmap_set_pmd(pmd_t *pmdp, void *p, int node,
 			       unsigned long addr, unsigned long next)
 {
@@ -1316,6 +1338,8 @@ int arch_add_memory(int nid, u64 start, u64 size,
 
 	if (can_set_direct_map())
 		flags |= NO_BLOCK_MAPPINGS | NO_CONT_MAPPINGS;
+	else if (should_pmd_linear_mapping())
+		flags |= NO_PUD_BLOCK_MAPPINGS | NO_CONT_MAPPINGS;
 
 	__create_pgd_mapping(swapper_pg_dir, start, __phys_to_virt(start),
 			     size, params->pgprot, __pgd_pgtable_alloc,
diff --git a/arch/arm64/mm/pfn_range_alloc.c b/arch/arm64/mm/pfn_range_alloc.c
new file mode 100644
index 000000000000..011b65bcbbaa
--- /dev/null
+++ b/arch/arm64/mm/pfn_range_alloc.c
@@ -0,0 +1,624 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Contiguous pfn range allocator
+ *
+ * Copyright (C) 2025 Huawei Limited.
+ */
+
+#define pr_fmt(fmt) "pfn_range_alloc: " fmt
+
+#include <linux/mm.h>
+#include <linux/memblock.h>
+#include <linux/page-isolation.h>
+#include <linux/set_memory.h>
+#include <trace/events/kmem.h>
+#include <linux/pagewalk.h>
+#include <linux/debugfs.h>
+#include <linux/seq_file.h>
+#include <linux/numa_remote.h>
+#include "internal.h"
+#include "../../../mm/internal.h"
+
+struct pmd_lm_range {
+	unsigned long start_pfn;
+	unsigned long end_pfn;
+	spinlock_t	lock;
+	unsigned long *bitmap;
+	unsigned long bitmap_maxno;
+};
+
+unsigned long contig_mem_pool_percent __ro_after_init;
+EXPORT_SYMBOL_GPL(contig_mem_pool_percent);
+static unsigned long nr_reserved_pages[MAX_NUMNODES] __initdata;
+static struct pmd_lm_range reserved_range[MAX_NUMNODES];
+DEFINE_STATIC_KEY_FALSE(pmd_mapping_initialized);
+static atomic_long_t num_poisoned_pfn __read_mostly = ATOMIC_LONG_INIT(0);
+
+static inline bool pmd_linear_mapping_enabled(void)
+{
+	return static_branch_unlikely(&pmd_mapping_initialized);
+}
+
+static __init int cmdline_parse_pmd_mapping(char *p)
+{
+	unsigned long percent;
+	char *endptr;
+
+	if (!p)
+		return -EINVAL;
+
+	percent = simple_strtoul(p, &endptr, 0);
+	if (*endptr != '%' || *(endptr + 1) != '\0')
+		return -EINVAL;
+
+	if (percent > 100)
+		return -EINVAL;
+
+	contig_mem_pool_percent = percent;
+
+	return 0;
+}
+early_param("pmd_mapping", cmdline_parse_pmd_mapping);
+
+static __init void calculate_node_nr_reserved_pages(void)
+{
+	unsigned long start_pfn, end_pfn;
+	int i, nid;
+
+	for_each_mem_pfn_range(i, MAX_NUMNODES, &start_pfn, &end_pfn, &nid)
+		nr_reserved_pages[nid] += end_pfn - start_pfn;
+
+	for_each_online_node(nid) {
+		nr_reserved_pages[nid] = nr_reserved_pages[nid] * contig_mem_pool_percent / 100;
+		nr_reserved_pages[nid] = ALIGN_DOWN(nr_reserved_pages[nid],
+							PUD_SIZE / PAGE_SIZE);
+	}
+}
+
+static __init unsigned long calculate_reserve_base(int nid)
+{
+	pg_data_t *pgdat = NODE_DATA(nid);
+	struct zone *zone;
+	unsigned long base = 0;
+
+#ifdef CONFIG_ZONE_DMA
+	zone = &pgdat->node_zones[ZONE_DMA];
+	if (managed_zone(zone))
+		base = max(base, PFN_PHYS(zone_end_pfn(zone)));
+#endif
+
+#ifdef CONFIG_ZONE_DMA32
+	zone = &pgdat->node_zones[ZONE_DMA32];
+	if (managed_zone(zone))
+		base = max(base, PFN_PHYS(zone_end_pfn(zone)));
+#endif
+
+	return base;
+}
+
+static __init int __get_suitable_reserved_range(int nid)
+{
+	unsigned long base, size, start;
+
+	base = calculate_reserve_base(nid);
+retry:
+	size = nr_reserved_pages[nid] * PAGE_SIZE;
+	start = memblock_alloc_range_nid(size, PUD_SIZE, base, 0, nid, true);
+	/*
+	 * If reservation fails, try to fallback to reserve
+	 * smaller size. Fallback is at PUD_SIZE granularity.
+	 */
+	if (!start) {
+		nr_reserved_pages[nid] -= PUD_SIZE / PAGE_SIZE;
+		if (!nr_reserved_pages[nid])
+			return -ENOMEM;
+		goto retry;
+	}
+
+	reserved_range[nid].start_pfn = PHYS_PFN(start);
+	reserved_range[nid].end_pfn = PHYS_PFN(start) + nr_reserved_pages[nid];
+
+	return 0;
+}
+
+static __init int get_suitable_reserved_range(void)
+{
+	bool restore_bottom_up = false;
+	unsigned long start, end;
+	bool resved = false;
+	int nid, ret;
+
+	if (memblock_bottom_up()) {
+		memblock_set_bottom_up(false);
+		restore_bottom_up = true;
+	}
+
+	calculate_node_nr_reserved_pages();
+	for_each_online_node(nid) {
+		if (!nr_reserved_pages[nid])
+			continue;
+
+		ret = __get_suitable_reserved_range(nid);
+		if (ret) {
+			pr_warn("reservation failed for node %d\n", nid);
+			continue;
+		}
+
+		start = PFN_PHYS(reserved_range[nid].start_pfn);
+		end = PFN_PHYS(reserved_range[nid].end_pfn);
+		pmd_mapping_reserved_remap(start, end);
+		resved = true;
+		pr_info("reserved %lu MiB on node %d\n", (end - start) / SZ_1M, nid);
+	}
+
+	if (restore_bottom_up)
+		memblock_set_bottom_up(true);
+
+	return resved;
+}
+
+static __init void put_suitable_reserved_range(void)
+{
+	unsigned long start, end;
+	int ret, nid;
+
+	for_each_online_node(nid) {
+		start = PFN_PHYS(reserved_range[nid].start_pfn);
+		end = PFN_PHYS(reserved_range[nid].end_pfn);
+
+		if (start == end)
+			continue;
+
+		ret = memblock_phys_free(start, end - start);
+		if (ret)
+			pr_warn("put reserved memory [%lx, %lx) failed(%d) for node %d\n",
+					start, end, ret, nid);
+	}
+}
+
+void __init pmd_mapping_reserve_and_remap(void)
+{
+	bool resved;
+
+	if (!contig_mem_pool_percent)
+		return;
+
+	if (should_pmd_linear_mapping())
+		goto out;
+
+	if (can_set_direct_map()) {
+		pr_info("linear mapping is mapped at PTE level, all memory can be borrowed\n");
+		goto out;
+	}
+
+	resved = get_suitable_reserved_range();
+	if (!resved)
+		return;
+
+	put_suitable_reserved_range();
+out:
+	static_branch_enable(&pmd_mapping_initialized);
+}
+
+static int __init activate_reserved_range(void)
+{
+	int nid;
+	unsigned long pfn, end_pfn;
+	unsigned long bitmap_maxno;
+
+	if (!pmd_linear_mapping_enabled())
+		return 0;
+
+	for_each_online_node(nid) {
+		pfn = reserved_range[nid].start_pfn;
+		end_pfn = reserved_range[nid].end_pfn;
+
+		if (pfn == end_pfn)
+			continue;
+
+		bitmap_maxno = (end_pfn - pfn) / (PFN_RANGE_ALLOC_SIZE / PAGE_SIZE);
+		reserved_range[nid].bitmap_maxno = bitmap_maxno;
+		reserved_range[nid].bitmap = bitmap_zalloc(bitmap_maxno, GFP_KERNEL);
+		if (!reserved_range[nid].bitmap) {
+			reserved_range[nid].start_pfn = 0;
+			reserved_range[nid].end_pfn = 0;
+			pr_warn("reserved_range %d fails to be initialized\n", nid);
+			continue;
+		}
+		spin_lock_init(&reserved_range[nid].lock);
+	}
+
+	return 0;
+}
+core_initcall(activate_reserved_range);
+
+struct folio *pfn_range_alloc(unsigned int nr_pages, int nid)
+{
+	unsigned long min_align = PFN_RANGE_ALLOC_NR_PAGES;
+	gfp_t gfp_mask = (GFP_KERNEL | __GFP_COMP) & ~__GFP_RECLAIM;
+	unsigned long start, bitmap_no, bitmap_count, mask, offset;
+	struct pmd_lm_range *mem_range;
+	struct folio *folio = ERR_PTR(-EINVAL);
+	unsigned long pfn;
+	int ret;
+
+	if (in_interrupt())
+		goto out;
+
+	if (nid < 0 || nid >= MAX_NUMNODES)
+		goto out;
+
+	if (!IS_ALIGNED(nr_pages, min_align))
+		goto out;
+
+	if (can_set_direct_map() || should_pmd_linear_mapping()) {
+		int order = ilog2(nr_pages);
+
+		folio = NULL;
+		gfp_mask |= __GFP_THISNODE;
+		if (nr_pages <= MAX_ORDER_NR_PAGES)
+			folio = __folio_alloc_node(gfp_mask | __GFP_NOWARN, order, nid);
+		if (!folio)
+			folio = folio_alloc_gigantic(order, gfp_mask, nid, NULL);
+		if (!folio)
+			folio = ERR_PTR(-ENOMEM);
+
+		goto out;
+	}
+
+	mem_range = &reserved_range[nid];
+	if (!mem_range->bitmap) {
+		folio = ERR_PTR(-ENOMEM);
+		goto out;
+	}
+
+	start = 0;
+	bitmap_count = nr_pages / min_align;
+	mask = bitmap_count - 1;
+	offset = (mem_range->start_pfn & (nr_pages - 1)) / min_align;
+	for (;;) {
+		spin_lock(&mem_range->lock);
+		bitmap_no = bitmap_find_next_zero_area_off(mem_range->bitmap,
+				mem_range->bitmap_maxno, start, bitmap_count, mask, offset);
+		if (bitmap_no >= mem_range->bitmap_maxno) {
+			spin_unlock(&mem_range->lock);
+			break;
+		}
+		bitmap_set(mem_range->bitmap, bitmap_no, bitmap_count);
+		spin_unlock(&mem_range->lock);
+		pfn = mem_range->start_pfn + bitmap_no * min_align;
+		ret = alloc_contig_range(pfn, pfn + nr_pages, MIGRATE_MOVABLE, gfp_mask);
+		if (!ret) {
+			folio = pfn_folio(pfn);
+			goto out;
+		}
+
+		spin_lock(&mem_range->lock);
+		bitmap_clear(mem_range->bitmap, bitmap_no, bitmap_count);
+		spin_unlock(&mem_range->lock);
+		start = bitmap_no + bitmap_count;
+	}
+
+	folio = ERR_PTR(-ENOMEM);
+out:
+	trace_pfn_range_alloc(folio, nr_pages, nid);
+	return folio;
+}
+EXPORT_SYMBOL_GPL(pfn_range_alloc);
+
+static void pfn_range_folio_dissolve(struct folio *folio)
+{
+	int nr_pages = folio_nr_pages(folio);
+	struct page *page;
+	int i;
+
+	VM_WARN_ON_FOLIO(folio_ref_count(folio) != 1, folio);
+
+	for (i = 1; i < nr_pages; i++) {
+		page = folio_page(folio, i);
+		page->flags &= ~PAGE_FLAGS_CHECK_AT_FREE;
+		page->mapping = NULL;
+		clear_compound_head(page);
+		set_page_refcounted(page);
+	}
+
+	__folio_clear_head(folio);
+
+	page = &folio->page;
+	for (i = 0; i < nr_pages; i++, page++) {
+		if (PageHWPoison(page)) {
+			atomic_long_inc(&num_poisoned_pfn);
+			continue;
+		}
+
+		__free_page(page);
+	}
+}
+
+static inline bool pfn_range_free_prepare(struct folio *folio)
+{
+	int nr_pages = folio_nr_pages(folio);
+	struct page *page = &folio->page;
+	int i;
+
+	for (i = 0; i < nr_pages; i++, page++)
+		if (PageHWPoison(page)) {
+			pfn_range_folio_dissolve(folio);
+			return false;
+		}
+
+	return true;
+}
+
+int pfn_range_free(struct folio *folio)
+{
+	struct pmd_lm_range *mem_range;
+	unsigned long start_pfn, end_pfn;
+	unsigned long bitmap_no, bitmap_count;
+	unsigned long nr_pages = folio_nr_pages(folio);
+	unsigned long min_align = PFN_RANGE_ALLOC_NR_PAGES;
+	int ret = 0;
+
+	if (in_interrupt()) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	if (!pfn_range_free_prepare(folio))
+		goto out;
+
+	if (can_set_direct_map() || should_pmd_linear_mapping()) {
+		folio_put(folio);
+		goto out;
+	}
+
+	mem_range = &reserved_range[folio_nid(folio)];
+	start_pfn = folio_pfn(folio);
+	end_pfn = start_pfn + nr_pages;
+
+	if (start_pfn < mem_range->start_pfn || end_pfn > mem_range->end_pfn) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	free_contig_range(start_pfn, nr_pages);
+	bitmap_no = (start_pfn - mem_range->start_pfn) / min_align;
+	bitmap_count = nr_pages / min_align;
+	spin_lock(&mem_range->lock);
+	bitmap_clear(mem_range->bitmap, bitmap_no, bitmap_count);
+	spin_unlock(&mem_range->lock);
+
+out:
+	trace_pfn_range_free(folio, ret);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(pfn_range_free);
+
+static inline int check_update_lm_arg(unsigned long start_pfn, unsigned long end_pfn)
+{
+	unsigned long start, end;
+	struct page *start_page;
+	int nid;
+
+	start_page = pfn_to_page(start_pfn);
+	nid = page_to_nid(start_page);
+	start = (unsigned long)page_to_virt(start_page);
+	end = start + (end_pfn - start_pfn) * PAGE_SIZE;
+	if ((start_pfn >= reserved_range[nid].start_pfn &&
+	    end_pfn <= reserved_range[nid].end_pfn)
+	    || should_pmd_linear_mapping()
+	    || can_set_direct_map()) {
+		if (!IS_ALIGNED(start, PFN_RANGE_ALLOC_SIZE) ||
+			!IS_ALIGNED(end, PFN_RANGE_ALLOC_SIZE)) {
+			return -EINVAL;
+		}
+	} else if (!IS_ALIGNED(start, PUD_SIZE) || !IS_ALIGNED(end, PUD_SIZE)) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int invalid_pud_entry(pud_t *pudp, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_invalid = (bool)walk->private;
+	pud_t pud;
+
+	pud = pudp_get(pudp);
+	if (pud_table(pud))
+		return 0;
+
+	if (set_invalid)
+		pud_val(pud) &= ~PTE_VALID;
+	else
+		pud_val(pud) |= PTE_VALID;
+	set_pud(pudp, pud);
+
+	return 0;
+}
+
+static int invalid_pmd_entry(pmd_t *pmdp, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_invalid = (bool)walk->private;
+	pmd_t pmd;
+
+	pmd = pmdp_get(pmdp);
+	if (pmd_table(pmd))
+		return 0;
+
+	if (set_invalid)
+		pmd_val(pmd) &= ~PTE_VALID;
+	else
+		pmd_val(pmd) |= PTE_VALID;
+	set_pmd(pmdp, pmd);
+
+	return 0;
+}
+
+static int invalid_pte_entry(pte_t *ptep, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_invalid = (bool)walk->private;
+	pte_t pte;
+
+	pte = ptep_get(ptep);
+
+	if (set_invalid)
+		pte_val(pte) &= ~PTE_VALID;
+	else
+		pte_val(pte) |= PTE_VALID;
+	set_pte(ptep, pte);
+
+	return 0;
+}
+
+static const struct mm_walk_ops invalid_ops = {
+	.pud_entry = invalid_pud_entry,
+	.pmd_entry = invalid_pmd_entry,
+	.pte_entry = invalid_pte_entry,
+};
+
+int set_linear_mapping_invalid(unsigned long start_pfn, unsigned long end_pfn,
+										bool set_invalid)
+{
+	unsigned long start, end;
+	int ret;
+
+	ret = check_update_lm_arg(start_pfn, end_pfn);
+	if (ret)
+		return ret;
+
+	start = (unsigned long)page_to_virt(pfn_to_page(start_pfn));
+	end = start + (end_pfn - start_pfn) * PAGE_SIZE;
+	mmap_write_lock(&init_mm);
+	walk_page_range_novma(&init_mm, start, end,
+				&invalid_ops, NULL, (void *)set_invalid);
+	mmap_write_unlock(&init_mm);
+	if (set_invalid)
+		flush_tlb_kernel_range(start, end);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(set_linear_mapping_invalid);
+
+static inline void update_entry_nc(unsigned long long *val, bool set_nc)
+{
+	*val &= ~PTE_ATTRINDX_MASK;
+	if (set_nc)
+		*val |= PTE_ATTRINDX(MT_NORMAL_NC);
+	else
+		*val |= PTE_ATTRINDX(MT_NORMAL_TAGGED);
+	*val |= PTE_VALID;
+}
+
+static int nc_pud_entry(pud_t *pudp, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_nc = (bool)walk->private;
+	pud_t pud;
+
+	if (pud_table(*pudp))
+		return 0;
+
+	pud = pudp_huge_get_and_clear(walk->mm, addr, pudp);
+	update_entry_nc(&pud_val(pud), set_nc);
+	set_pud(pudp, pud);
+
+	return 0;
+}
+
+static int nc_pmd_entry(pmd_t *pmdp, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_nc = (bool)walk->private;
+	pmd_t pmd;
+
+	if (pmd_table(*pmdp))
+		return 0;
+
+	pmd = pmdp_huge_get_and_clear(walk->mm, addr, pmdp);
+	update_entry_nc(&pmd_val(pmd), set_nc);
+	set_pmd(pmdp, pmd);
+
+	return 0;
+}
+
+static int nc_pte_entry(pte_t *ptep, unsigned long addr,
+				unsigned long next, struct mm_walk *walk)
+{
+	bool set_nc = (bool)walk->private;
+	pte_t pte;
+
+	pte = ptep_get_and_clear(walk->mm, addr, ptep);
+	update_entry_nc(&pte_val(pte), set_nc);
+	set_pte(ptep, pte);
+
+	return 0;
+}
+
+static const struct mm_walk_ops nc_ops = {
+	.pud_entry = nc_pud_entry,
+	.pmd_entry = nc_pmd_entry,
+	.pte_entry = nc_pte_entry,
+};
+
+int set_linear_mapping_nc(unsigned long start_pfn, unsigned long end_pfn, bool set_nc)
+{
+	unsigned long start, end;
+	int ret;
+
+	start = (unsigned long)page_to_virt(pfn_to_page(start_pfn));
+	end = start + (end_pfn - start_pfn) * PAGE_SIZE;
+	ret = check_update_lm_arg(start_pfn, end_pfn);
+	if (ret)
+		return ret;
+
+	mmap_write_lock(&init_mm);
+	walk_page_range_novma(&init_mm, start, end,
+				&invalid_ops, NULL, (void *)true);
+	flush_tlb_kernel_range(start, end);
+	walk_page_range_novma(&init_mm, start, end,
+				&nc_ops, NULL, (void *)set_nc);
+	mmap_write_unlock(&init_mm);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(set_linear_mapping_nc);
+
+#ifdef CONFIG_DEBUG_FS
+static int reserved_range_show(struct seq_file *m, void *v)
+{
+	int nid;
+
+	for_each_online_node(nid) {
+		if (numa_is_remote_node(nid))
+			continue;
+
+		seq_printf(m, "%d, %llx-%llx\n", nid,
+				PFN_PHYS(reserved_range[nid].start_pfn),
+				PFN_PHYS(reserved_range[nid].end_pfn));
+	}
+
+	seq_printf(m, "\nHardwareCorrupted: %lu kB\n",
+		   atomic_long_read(&num_poisoned_pfn) << (PAGE_SHIFT - 10));
+
+	return 0;
+}
+DEFINE_SHOW_ATTRIBUTE(reserved_range);
+
+static int __init reserved_range_debug_init(void)
+{
+	if (!pmd_linear_mapping_enabled())
+		return 0;
+
+	if (can_set_direct_map() || should_pmd_linear_mapping())
+		return 0;
+
+	debugfs_create_file("pmd_mapping_reserved_range", 0400, NULL,
+			    NULL, &reserved_range_fops);
+	return 0;
+}
+late_initcall(reserved_range_debug_init);
+#endif
diff --git a/include/linux/hugetlb.h b/include/linux/hugetlb.h
index aef07cd9f313..1acb5155a235 100644
--- a/include/linux/hugetlb.h
+++ b/include/linux/hugetlb.h
@@ -1339,4 +1339,24 @@ hugetlb_walk(struct vm_area_struct *vma, unsigned long addr, unsigned long sz)
 	return huge_pte_offset(vma->vm_mm, addr, sz);
 }
 
+struct folio *alloc_hugetlb_folio_size(int nid, unsigned long size);
+#ifdef CONFIG_PFN_RANGE_ALLOC
+struct folio *hugetlb_pool_alloc(int nid);
+int hugetlb_pool_free(struct folio *folio);
+struct folio *hugetlb_pool_alloc_size(int nid, unsigned long size);
+#else
+static inline struct folio *hugetlb_pool_alloc(int nid)
+{
+	return ERR_PTR(-EINVAL);
+}
+static inline int hugetlb_pool_free(struct folio *folio)
+{
+	return -EINVAL;
+}
+static inline struct folio *hugetlb_pool_alloc_size(int nid, unsigned long size)
+{
+	return ERR_PTR(-EINVAL);
+}
+#endif
+
 #endif /* _LINUX_HUGETLB_H */
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 8fe90c1a7712..a15d037ec05a 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -4266,4 +4266,36 @@ static inline bool pfn_is_unaccepted_memory(unsigned long pfn)
 	return range_contains_unaccepted_memory(paddr, paddr + PAGE_SIZE);
 }
 
+#ifdef CONFIG_PFN_RANGE_ALLOC
+#define PFN_RANGE_ALLOC_SIZE PMD_SIZE
+#define PFN_RANGE_ALLOC_ORDER PMD_ORDER
+#define PFN_RANGE_ALLOC_NR_PAGES (1 << PFN_RANGE_ALLOC_ORDER)
+
+extern unsigned long contig_mem_pool_percent;
+struct folio *pfn_range_alloc(unsigned int nr_pages, int nid);
+int pfn_range_free(struct folio *folio);
+int set_linear_mapping_nc(unsigned long start_pfn, unsigned long end_pfn, bool set_nc);
+int set_linear_mapping_invalid(unsigned long start_pfn, unsigned long end_pfn,
+										bool set_invalid);
+#else
+static inline struct folio *pfn_range_alloc(unsigned int nr_pages, int nid)
+{
+	return ERR_PTR(-EINVAL);
+}
+static inline int pfn_range_free(struct folio *folio)
+{
+	return -EINVAL;
+}
+static inline
+int set_linear_mapping_nc(unsigned long start_pfn, unsigned long end_pfn, bool set_nc)
+{
+	return -EINVAL;
+}
+static inline
+int set_linear_mapping_invalid(unsigned long start_pfn, unsigned long end_pfn,
+										bool set_invalid)
+{
+	return -EINVAL;
+}
+#endif
 #endif /* _LINUX_MM_H */
diff --git a/include/trace/events/kmem.h b/include/trace/events/kmem.h
index 6e62cc64cd92..6d45a12e717f 100644
--- a/include/trace/events/kmem.h
+++ b/include/trace/events/kmem.h
@@ -409,6 +409,124 @@ TRACE_EVENT(rss_stat,
 		__print_symbolic(__entry->member, TRACE_MM_PAGES),
 		__entry->size)
 	);
+
+#ifdef CONFIG_PFN_RANGE_ALLOC
+TRACE_EVENT(pfn_range_alloc,
+
+	TP_PROTO(struct folio *folio, unsigned int nr_pages,
+			int nid),
+
+	TP_ARGS(folio, nr_pages, nid),
+
+	TP_STRUCT__entry(
+		__field(struct folio*, folio)
+		__field(unsigned int, nr_pages)
+		__field(int, nid)
+	),
+
+	TP_fast_assign(
+		__entry->folio = folio;
+		__entry->nr_pages = nr_pages;
+		__entry->nid = nid;
+	),
+
+	TP_printk("folio=%p pfn=0x%lx err=%ld nr_pages=%u nid=%d",
+		!IS_ERR(__entry->folio) ? __entry->folio : NULL,
+		!IS_ERR(__entry->folio) ? folio_pfn(__entry->folio) : 0,
+		!IS_ERR(__entry->folio) ? 0 : PTR_ERR(__entry->folio),
+		__entry->nr_pages,
+		__entry->nid)
+);
+
+TRACE_EVENT(pfn_range_free,
+
+	TP_PROTO(struct folio *folio, int ret),
+
+	TP_ARGS(folio, ret),
+
+	TP_STRUCT__entry(
+		__field(struct folio*, folio)
+		__field(int, ret)
+	),
+
+	TP_fast_assign(
+		__entry->folio = folio;
+		__entry->ret = ret;
+	),
+
+	TP_printk("folio=%p pfn=0x%lx ret=%d",
+		__entry->folio, folio_pfn(__entry->folio), __entry->ret)
+);
+
+TRACE_EVENT(hugetlb_pool_alloc,
+
+	TP_PROTO(struct folio *folio, int nid),
+
+	TP_ARGS(folio, nid),
+
+	TP_STRUCT__entry(
+		__field(struct folio*, folio)
+		__field(int, nid)
+	),
+
+	TP_fast_assign(
+		__entry->folio = folio;
+		__entry->nid = nid;
+	),
+
+	TP_printk("folio=%p pfn=0x%lx err=%ld nid=%d",
+		!IS_ERR(__entry->folio) ? __entry->folio : NULL,
+		!IS_ERR(__entry->folio) ? folio_pfn(__entry->folio) : 0,
+		!IS_ERR(__entry->folio) ? 0 : PTR_ERR(__entry->folio),
+		__entry->nid)
+);
+
+TRACE_EVENT(hugetlb_pool_free,
+
+	TP_PROTO(struct folio *folio, int ret),
+
+	TP_ARGS(folio, ret),
+
+	TP_STRUCT__entry(
+		__field(struct folio*, folio)
+		__field(int, ret)
+	),
+
+	TP_fast_assign(
+		__entry->folio = folio;
+		__entry->ret = ret;
+	),
+
+	TP_printk("folio=%p pfn=0x%lx ret=%d",
+		__entry->folio, folio_pfn(__entry->folio), __entry->ret)
+);
+
+TRACE_EVENT(hugetlb_pool_alloc_size,
+
+	TP_PROTO(struct folio *folio, int nid, unsigned long size),
+
+	TP_ARGS(folio, nid, size),
+
+	TP_STRUCT__entry(
+		__field(struct folio*, folio)
+		__field(int, nid)
+		__field(unsigned long, size)
+	),
+
+	TP_fast_assign(
+		__entry->folio = folio;
+		__entry->nid = nid;
+		__entry->size = size;
+	),
+
+	TP_printk("folio=%p pfn=0x%lx err=%ld nid=%d size=0x%lx",
+		!IS_ERR(__entry->folio) ? __entry->folio : NULL,
+		!IS_ERR(__entry->folio) ? folio_pfn(__entry->folio) : 0,
+		!IS_ERR(__entry->folio) ? 0 : PTR_ERR(__entry->folio),
+		__entry->nid, __entry->size)
+);
+
+#endif
 #endif /* _TRACE_KMEM_H */
 
 /* This part must be outside protection */
diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index af989f84dd8a..789c3f2c8209 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -48,6 +48,7 @@
 #include <linux/node.h>
 #include <linux/page_owner.h>
 #include <linux/numa_remote.h>
+#include <trace/events/kmem.h>
 #include "internal.h"
 #include "hugetlb_vmemmap.h"
 
@@ -8079,3 +8080,83 @@ static void __init hugetlb_cma_check(void)
 }
 
 #endif /* CONFIG_CMA */
+
+struct folio *alloc_hugetlb_folio_size(int nid, unsigned long size)
+{
+	gfp_t gfp_mask;
+	struct hstate *h;
+	nodemask_t nodemask;
+	unsigned long flags;
+	struct folio *folio = NULL;
+
+	h = size_to_hstate(size);
+	if (!h)
+		return NULL;
+
+	nodes_clear(nodemask);
+	node_set(nid, nodemask);
+	gfp_mask = htlb_alloc_mask(h);
+
+	spin_lock_irqsave(&hugetlb_lock, flags);
+	if (h->free_huge_pages - h->resv_huge_pages > 0)
+		folio = dequeue_hugetlb_folio_nodemask(h, gfp_mask, nid, &nodemask);
+	spin_unlock_irqrestore(&hugetlb_lock, flags);
+
+	return folio;
+}
+EXPORT_SYMBOL(alloc_hugetlb_folio_size);
+
+#ifdef CONFIG_PFN_RANGE_ALLOC
+struct folio *hugetlb_pool_alloc(int nid)
+{
+	struct folio *folio = ERR_PTR(-EINVAL);
+
+	if (nid < 0 || nid >= MAX_NUMNODES)
+		goto out;
+
+	folio = alloc_hugetlb_folio_size(nid, PFN_RANGE_ALLOC_SIZE);
+	if (!folio)
+		folio = ERR_PTR(-ENOMEM);
+
+out:
+	trace_hugetlb_pool_alloc(folio, nid);
+	return folio;
+}
+EXPORT_SYMBOL_GPL(hugetlb_pool_alloc);
+
+int hugetlb_pool_free(struct folio *folio)
+{
+	int ret = -EINVAL;
+
+	if (!folio_test_hugetlb(folio))
+		goto out;
+
+	ret = 0;
+	folio_put(folio);
+out:
+	trace_hugetlb_pool_free(folio, ret);
+	return ret;
+}
+EXPORT_SYMBOL_GPL(hugetlb_pool_free);
+
+struct folio *hugetlb_pool_alloc_size(int nid, unsigned long size)
+{
+	struct folio *folio = ERR_PTR(-EINVAL);
+
+	if (nid < 0 || nid >= MAX_NUMNODES)
+		goto out;
+
+	if ((size != PMD_SIZE) && (size != PUD_SIZE))
+		goto out;
+
+	folio = alloc_hugetlb_folio_size(nid, size);
+	if (!folio)
+		folio = ERR_PTR(-ENOMEM);
+
+out:
+	trace_hugetlb_pool_alloc_size(folio, nid, size);
+	return folio;
+}
+EXPORT_SYMBOL_GPL(hugetlb_pool_alloc_size);
+#endif
+
diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index c02f972d79e1..f4b101d18526 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -6760,6 +6760,35 @@ static void split_free_pages(struct list_head *list, gfp_t gfp_mask)
 	}
 }
 
+static void split_pages_to_order0(struct list_head *list)
+{
+	int order;
+
+	for (order = 1; order < NR_PAGE_ORDERS; order++) {
+		struct page *page, *next;
+		int nr_pages = 1 << order;
+
+		list_for_each_entry_safe(page, next, &list[order], lru) {
+			int i;
+
+			list_del(&page->lru);
+			for (i = 0; i < nr_pages; i++)
+				list_add_tail(&page[i].lru, &list[0]);
+		}
+	}
+}
+
+static void free_pfn_range(unsigned long start, unsigned long end, gfp_t gfp_mask)
+{
+	struct page *page;
+	unsigned long i;
+
+	page = pfn_to_page(start);
+	for (i = 0; i < end - start; ++i, ++page)
+		post_alloc_hook(page, 0, gfp_mask);
+	free_contig_range(start, end - start);
+}
+
 static int __alloc_contig_verify_gfp_mask(gfp_t gfp_mask, gfp_t *gfp_cc_mask)
 {
 	const gfp_t reclaim_mask = __GFP_IO | __GFP_FS | __GFP_RECLAIM;
@@ -6820,6 +6849,7 @@ static int __alloc_contig_verify_gfp_mask(gfp_t gfp_mask, gfp_t *gfp_cc_mask)
 int alloc_contig_range(unsigned long start, unsigned long end,
 		       unsigned migratetype, gfp_t gfp_mask)
 {
+	int range_order = ilog2(end - start);
 	unsigned long outer_start, outer_end;
 	int ret = 0;
 
@@ -6838,6 +6868,14 @@ int alloc_contig_range(unsigned long start, unsigned long end,
 	if (__alloc_contig_verify_gfp_mask(gfp_mask, (gfp_t *)&cc.gfp_mask))
 		return -EINVAL;
 
+	/* __GFP_COMP may only be used for certain aligned+sized ranges. */
+	if ((gfp_mask & __GFP_COMP) &&
+		(!is_power_of_2(end - start) || !IS_ALIGNED(start, 1 << range_order))) {
+		WARN_ONCE(true, "PFN range: requested [%lu, %lu) is not suitable for __GFP_COMP\n",
+				start, end);
+		return -EINVAL;
+	}
+
 	/*
 	 * What we do here is we mark all pageblocks in range as
 	 * MIGRATE_ISOLATE.  Because pageblock and max order pages may
@@ -6911,6 +6949,11 @@ int alloc_contig_range(unsigned long start, unsigned long end,
 		goto done;
 	}
 
+	/*
+	 * With __GFP_COMP and the requested order < MAX_PAGE_ORDER,
+	 * isolated free pages can have higher order than the requested
+	 * one. Use split_free_pages() to free out of range pages.
+	 */
 	if (!(gfp_mask & __GFP_COMP)) {
 		split_free_pages(cc.freepages, gfp_mask);
 
@@ -6919,16 +6962,20 @@ int alloc_contig_range(unsigned long start, unsigned long end,
 			free_contig_range(outer_start, start - outer_start);
 		if (end != outer_end)
 			free_contig_range(end, outer_end - end);
-	} else if (start == outer_start && end == outer_end && is_power_of_2(end - start)) {
+	} else {
 		struct page *head = pfn_to_page(start);
-		int order = ilog2(end - start);
 
-		check_new_pages(head, order);
-		prep_new_page(head, order, gfp_mask, 0);
-	} else {
-		ret = -EINVAL;
-		WARN(true, "PFN range: requested [%lu, %lu), allocated [%lu, %lu)\n",
-		     start, end, outer_start, outer_end);
+		if ((outer_start != start) || (end != outer_end)) {
+			split_pages_to_order0(cc.freepages);
+			if (start != outer_start)
+				free_pfn_range(outer_start, start, gfp_mask);
+
+			if (end != outer_end)
+				free_pfn_range(end, outer_end, gfp_mask);
+		}
+
+		check_new_pages(head, range_order);
+		prep_new_page(head, range_order, gfp_mask, 0);
 	}
 done:
 	undo_isolate_page_range(start, end, migratetype);
-- 
2.43.0

